### ISSUE TITLE
Only one pointlight at a time possible

### ISSUE THREAD

**Issue by Simonrazer** [armory3d/armory#1772]
Links to #1140 #1429, reposting because these issues don't have the error posted and don't get the attention they really need.

**Description**
When using more than one pointlamp this error occurs: 
When using Deferred rendering:
```
Compiling shader 20 of 20 (World_World.vert.glsl).
C:\Users\SK\Desktop\Armory\blends\test\build_test\debug\temp\deferred_light.frag.hlsl(152,20-84): warning X3570: gradient instruction used in a loop with varying iteration, attempting to unroll the loop
C:\Users\SK\Desktop\Armory\blends\test\build_test\debug\temp\deferred_light.frag.hlsl(5,14-30): error X4500: overlapping register semantics not yet implemented 's1'
C:\Users\SK\Desktop\Armory\blends\test\build_test\debug\temp\deferred_light.frag.hlsl(7,14-30): error X4500: overlapping register semantics not yet implemented 's2'
C:\Users\SK\Desktop\Armory\blends\test\build_test\debug\temp\deferred_light.frag.hlsl(9,14-30): error X4500: overlapping register semantics not yet implemented 's3'
C:\Users\SK\Desktop\Armory\blends\test\build_test\debug\temp\deferred_light.frag.hlsl(4,19-26): error X4500: overlapping register semantics not yet implemented 't1'
C:\Users\SK\Desktop\Armory\blends\test\build_test\debug\temp\deferred_light.frag.hlsl(6,19-26): error X4500: overlapping register semantics not yet implemented 't2'
C:\Users\SK\Desktop\Armory\blends\test\build_test\debug\temp\deferred_light.frag.hlsl(8,19-26): error X4500: overlapping register semantics not yet implemented 't3'

Compiling shader 8 of 20 (deferred_light.frag.glsl) failed:
Shader compiler error.
(node:8276) UnhandledPromiseRejectionWarning: Shader compiler error.
(node:8276) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
(node:8276) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
ERROR: Build failed, check console
```
When using Forward rendering:
```
Compiling shader 19 of 19 (World_World.vert.glsl).
C:\Users\SK\Desktop\Armory\blends\test\build_test\debug\temp\armdefault_mesh.frag.hlsl(125,20-84): warning X3570: gradient instruction used in a loop with varying iteration, attempting to unroll the loop
C:\Users\SK\Desktop\Armory\blends\test\build_test\debug\temp\armdefault_mesh.frag.hlsl(8,19-30): error X4500: overlapping register semantics not yet implemented 't1'

Compiling shader 1 of 19 (armdefault_mesh.frag.glsl) failed:
Shader compiler error.
(node:12752) UnhandledPromiseRejectionWarning: Shader compiler error.
(node:12752) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
(node:12752) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
C:\Users\SK\Desktop\Armory\blends\test\build_test\debug\temp\Material_mesh.frag.hlsl(125,20-84): warning X3570: gradient instruction used in a loop with varying iteration, attempting to unroll the loop
C:\Users\SK\Desktop\Armory\blends\test\build_test\debug\temp\Material_mesh.frag.hlsl(8,19-30): error X4500: overlapping register semantics not yet implemented 't1'

Compiling shader 8 of 19 (Material_mesh.frag.glsl) failed:
Shader compiler error.
ERROR: Build failed, check console
```
This makes it impossible to build bigger scenes in Armory.

**To Reproduce**
Place more than one pointlamp

**System**
Blender: 2.82/2.83
Armory: latest git
OS: Win 10
Graphics card: Gtx 1060

[lampbug.zip](https://github.com/armory3d/armory/files/4941367/lampbug.zip)

**Comment by knowledgenude** [comment]
Here it is working i guess 
![Screenshot_20200718_113227](https://user-images.githubusercontent.com/63247726/87854769-88b9c580-c8ea-11ea-8b2f-1e1312405486.jpg)

[untitled.zip](https://github.com/armory3d/armory/files/4941819/untitled.zip)

**Comment by Simonrazer** [comment]
Tried your file, not working for me (and probably many others too, judging by the similar issues I missed to link).
Multiple lights used to work some months ago or so, are you sure you have a up-to-date version?

**Comment by knowledgenude** [comment]
I use Git version. I also have this problem in Windows but not in Linux. Did you already tryed with a Windows or Linux Krom export?

**Comment by Simonrazer** [comment]
I don't have a Linux machine. A Windows Krom export does result in the deferred error. If I set the Play mode to Browser in works though.

**Comment by knowledgenude** [comment]
Then the problem is happening in Windows Krom export, as it occurs in Krom and Kromx (not tested in C). I will test other builds here and see what happens

**Comment by MoritzBrueckner** [comment]
There was a small discussion about this in the discord channel:
https://discordapp.com/channels/486771218599510021/486771218599510024/713166567134724118

**Summary:**
Without knowing anything about HLSL (that's where the error occurs), I assume the error lays within the declaration of uniforms (if it's even called like that in HLSL too).

Instead of this:
```hlsl
SamplerComparisonState _shadowMapSpot_sampler[4] : register(s0);
SamplerComparisonState _shadowMapPoint_sampler[4] : register(s1);
```

The generated output should probably look like this (source: https://stackoverflow.com/a/32479831/9985959):
```hlsl
SamplerComparisonState _shadowMapSpot_sampler[4] : register(s0);
SamplerComparisonState _shadowMapPoint_sampler[4] : register(s4); // <-- s4 instead of s1 because we the array has length 4
```

I'm unfortunately not able to help here, maybe it's even a Kha issue (because of the glsl -> hlsl conversion), I don't know.

**Comment by knowledgenude** [comment]
I can't test with DirectX now, but in OpenGL it is working fine. I only get an issue trying to use Vulkan API: 
![Screenshot_20200721_225953](https://user-images.githubusercontent.com/63247726/88125710-415f5d80-cba6-11ea-8adb-11bf41819aa5.jpg)

My GPU supports Vulkan, maybe it is missing some driver. Can someone please test if Vulkan is working or not in Armory with the example above?

**Comment by Simonrazer** [comment]
I tried to build Kromx for Vulkan but it fails. As for OpenGL, it build a Krom.exe without any errors, but when I actually try to build the Game with Armory and then launch it nothing happens. When I try to open it via commandline no messages are there, it is just instantly done. That might be because I did stuff wrong, but I am don't know what that could be. I used these [instructions](https://github.com/armory3d/Kromx/blob/master/README.md) to build Kromx, I just replaced the "direct3d11" with "opengl" or "vulkan".

**Comment by e1e5en-gd** [comment]
When adding a second light to the scene, various errors occur with the `deferred_light.frag.hlsl` shader:
1. Point and Spot (Spot and Spot)
![use-case-1](https://user-images.githubusercontent.com/7114353/97398964-32065380-18fd-11eb-9352-af2771c60938.jpg)
```
...\two_lights\build_error_two_lights\debug\temp\deferred_light.frag.hlsl(146,20-75): warning X3570: gradient instruction used in a loop with varying iteration, attempting to unroll the loop
...\two_lights\build_error_two_lights\debug\temp\deferred_light.frag.hlsl(290,5): warning X4000: use of potentially uninitialized variable (sampleLight)
...\two_lights\build_error_two_lights\debug\temp\deferred_light.frag.hlsl(8,24-46): error X4500: overlapping register semantics not yet implemented 's1'
...\two_lights\build_error_two_lights\debug\temp\deferred_light.frag.hlsl(8,24-46): error X4500: overlapping register semantics not yet implemented 's2'
...\two_lights\build_error_two_lights\debug\temp\deferred_light.frag.hlsl(8,24-46): error X4500: overlapping register semantics not yet implemented 's3'
...\two_lights\build_error_two_lights\debug\temp\deferred_light.frag.hlsl(11,14-30): error X4500: overlapping register semantics not yet implemented 's2'
...\two_lights\build_error_two_lights\debug\temp\deferred_light.frag.hlsl(13,14-30): error X4500: overlapping register semantics not yet implemented 's3'
...\two_lights\build_error_two_lights\debug\temp\deferred_light.frag.hlsl(15,14-30): error X4500: overlapping register semantics not yet implemented 's4'
...\two_lights\build_error_two_lights\debug\temp\deferred_light.frag.hlsl(7,21-34): error X4500: overlapping register semantics not yet implemented 't1'
...\two_lights\build_error_two_lights\debug\temp\deferred_light.frag.hlsl(7,21-34): error X4500: overlapping register semantics not yet implemented 't2'
...\two_lights\build_error_two_lights\debug\temp\deferred_light.frag.hlsl(7,21-34): error X4500: overlapping register semantics not yet implemented 't3'
...\two_lights\build_error_two_lights\debug\temp\deferred_light.frag.hlsl(10,19-26): error X4500: overlapping register semantics not yet implemented 't2'
...\two_lights\build_error_two_lights\debug\temp\deferred_light.frag.hlsl(12,19-26): error X4500: overlapping register semantics not yet implemented 't3'
...\two_lights\build_error_two_lights\debug\temp\deferred_light.frag.hlsl(14,19-26): error X4500: overlapping register semantics not yet implemented 't4'

Compiling shader 4 of 19 (deferred_light.frag.glsl) failed:
Shader compiler error.
(node:4596) UnhandledPromiseRejectionWarning: Shader compiler error.
(node:4596) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
(node:4596) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

2. Point and Point
![use-case-2](https://user-images.githubusercontent.com/7114353/97398984-3d597f00-18fd-11eb-8ed6-494361e4b517.jpg)
```
...\two_lights\build_error_two_lights\debug\temp\deferred_light.frag.hlsl(152,20-84): warning X3570: gradient instruction used in a loop with varying iteration, attempting to unroll the loop
...\two_lights\build_error_two_lights\debug\temp\deferred_light.frag.hlsl(5,14-30): error X4500: overlapping register semantics not yet implemented 's1'
...\two_lights\build_error_two_lights\debug\temp\deferred_light.frag.hlsl(7,14-30): error X4500: overlapping register semantics not yet implemented 's2'
...\two_lights\build_error_two_lights\debug\temp\deferred_light.frag.hlsl(9,14-30): error X4500: overlapping register semantics not yet implemented 's3'
...\two_lights\build_error_two_lights\debug\temp\deferred_light.frag.hlsl(4,19-26): error X4500: overlapping register semantics not yet implemented 't1'
...\two_lights\build_error_two_lights\debug\temp\deferred_light.frag.hlsl(6,19-26): error X4500: overlapping register semantics not yet implemented 't2'
...\two_lights\build_error_two_lights\debug\temp\deferred_light.frag.hlsl(8,19-26): error X4500: overlapping register semantics not yet implemented 't3'

Compiling shader 4 of 19 (deferred_light.frag.glsl) failed:
Shader compiler error.
(node:9004) UnhandledPromiseRejectionWarning: Shader compiler error.
(node:9004) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
(node:9004) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

3. Point and Area
![use-case-3](https://user-images.githubusercontent.com/7114353/97399006-4a766e00-18fd-11eb-85c2-b3ffd38a6daf.jpg)
```
...\two_lights\build_error_two_lights\debug\temp\deferred_light.frag.hlsl(361,20-75): warning X3570: gradient instruction used in a loop with varying iteration, attempting to unroll the loop
...\two_lights\build_error_two_lights\debug\temp\deferred_light.frag.hlsl(19,14-30): error X4500: overlapping register semantics not yet implemented 's4'
...\two_lights\build_error_two_lights\debug\temp\deferred_light.frag.hlsl(21,14-30): error X4500: overlapping register semantics not yet implemented 's5'
...\two_lights\build_error_two_lights\debug\temp\deferred_light.frag.hlsl(18,19-26): error X4500: overlapping register semantics not yet implemented 't4'
...\two_lights\build_error_two_lights\debug\temp\deferred_light.frag.hlsl(20,19-26): error X4500: overlapping register semantics not yet implemented 't5'

Compiling shader 4 of 19 (deferred_light.frag.glsl) failed:
Shader compiler error.
(node:9916) UnhandledPromiseRejectionWarning: Shader compiler error.
(node:9916) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
(node:9916) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

When using `Sun` light source with other sources - no errors.
In this case, errors occur even if the light sources are in **different scenes** of the same project.

**System**
Blender: 2.83.8
Armory: git, master (9e8117e)
OS: Windows 10 64-bit
Graphics card: NVIDIA GeForce 930M

**Test File**
[error_two_lights.zip](https://github.com/armory3d/armory/files/5449972/error_two_lights.zip)

**Comment by MoritzBrueckner** [comment]
There is a workaround now, thanks to #2102. When the new `Shadow Map Atlasing` option is enabled, multiple lights are now also working on DirectX.

**Comment by MoritzBrueckner** [comment]
Found a minimal reproducible example file for Krafix, so it's a Krafix bug: https://github.com/Kode/krafix/issues/88

**Comment by MoritzBrueckner** [comment]
Fixed in SDK 22.07 :)
